R/deploymentTarget.R

```diff
@@ -1,93 +1,241 @@
+  if (!is.null(appId)) {
+    return(deploymentTargetFromAppId(
+      recordPath = recordPath,
+      appId = appId,
+      appName = appName,
+      appTitle = appTitle,
+      envVars = envVars,
+      account = account,
+      server = server,
+      error_call = error_call
+    ))
+  }
+
+  if (!is.null(appName)) {
+    return(deploymentTargetFromAppName(
+      recordPath = recordPath,
+      appName = appName,
+      appTitle = appTitle,
+      envVars = envVars,
+      account = account,
+      server = server,
+      forceUpdate = forceUpdate,
+      error_call = error_call
+    ))
+  }
+
+  # No identifying appId or appName.
+
+  # When there are existing deployments, ask the user to select one and use
+  # it. Only deployments associated with locally configured account+server
+  # combinations are considered.
+  allDeployments <- deployments(
     appPath = recordPath,
-    nameFilter = appName,
     accountFilter = account,
     serverFilter = server
   )
+  if (nrow(allDeployments) > 0) {
+    deployment <- disambiguateDeployments(allDeployments, error_call = error_call)
+    deployment <- updateDeploymentTarget(deployment, appTitle, envVars)
+    accountDetails <- accountInfo(deployment$account, deployment$server)
+    return(list(
+      accountDetails = accountDetails,
+      deployment = deployment
+    ))
+  }
 
-  if (nrow(appDeployments) == 0) {
-    fullAccount <- findAccount(account, server)
-    if (is.null(appName)) {
-      appName <- defaultAppName(recordPath, fullAccount$server)
-    } else {
-      check_string(appName, call = error_call)
-    }
+  # Otherwise, identify a target account (given just one available or prompted
+  # by the user), generate a name, and locate the deployment.
+  accountDetails <- accountInfo(account, server)
+  appName <- generateAppName(appTitle, recordPath, accountDetails$name, unique = FALSE)
+  return(deploymentTargetFromAppName(
+    recordPath = recordPath,
+    appName = appName,
+    appTitle = appTitle,
+    envVars = envVars,
+    account = accountDetails$name,
+    server = accountDetails$server,
+    forceUpdate = forceUpdate,
+    error_call = error_call
+  ))
+}
 
-    appId <- NULL
-    if (!isPositCloudServer(fullAccount$server)) {
-      # Have we previously deployed elsewhere? We can't do this on cloud
-      # because it assigns random app names (see #808 for details).
-      existing <- applications(fullAccount$name, fullAccount$server)
-      if (appName %in% existing$name) {
-        thisApp <- existing[appName == existing$name, , drop = FALSE]
-        uniqueName <- findUnique(appName, existing$name)
-
-        if (shouldUpdateApp(thisApp, uniqueName, forceUpdate)) {
-          appId <- thisApp$id
-          appName <- thisApp$name
-        } else {
-          appName <- uniqueName
-        }
-      }
-    }
+# Discover the deployment target given appId.
+#
+# When appId is provided, all other information is secondary. An appId is an indication from the
```

Comment:
I found this comment very useful for understand the problem and where I went wrong previously.

Reply:
Thanks. It took _so_ much code-reading and reasoning about the previous workflow to see it. I'm hoping that being much more explicit will help us in the future.